packages/core/src/client.ts

```diff
@@ -1,28 +1,105 @@
-const { add, clear, merge } = require('./lib/feature-flag-delegate')
+import configSchema from './config'
+import Event from './event'
+import Breadcrumb from './breadcrumb'
+import Session from './session'
+import map from './lib/es-utils/map'
+import includes from './lib/es-utils/includes'
+import filter from './lib/es-utils/filter'
+import reduce from './lib/es-utils/reduce'
+import keys from './lib/es-utils/keys'
+import assign from './lib/es-utils/assign'
+import runCallbacks from './lib/callback-runner'
+import metadataDelegate from './lib/metadata-delegate'
+import runSyncCallbacks from './lib/sync-callback-runner'
+import BREADCRUMB_TYPES from './lib/breadcrumb-types'
+import { add, clear, merge } from './lib/feature-flag-delegate'
+import { App, BreadcrumbType, Config, Device, FeatureFlag, OnBreadcrumbCallback, OnErrorCallback, OnSessionCallback, Plugin, User } from './common'
 
 const noop = () => {}
 
-class Client {
-  constructor (configuration, schema = config.schema, internalPlugins = [], notifier) {
+interface LoggerConfig {
+  debug: (msg: any) => void
+  info: (msg: any) => void
+  warn: (msg: any) => void
+  error: (msg: any, err?: unknown) => void
+}
+
+interface Notifier {
+  name: string
+  version: string
+  url: string
+}
+
+interface EventDeliveryPayload {
+  apiKey: string
+  notifier: Notifier
+  events: Event[]
+}
+
+interface SessionDeliveryPayload {
+  notifier?: Notifier
+  device?: Device
+  app?: App
+  sessions?: Array<{
+    id: string
+    startedAt: Date
+    user?: User
+  }>
+}
+
+interface Delivery {
+  sendEvent(payload: EventDeliveryPayload, cb: (err?: Error | null) => void): void
+  sendSession(session: SessionDeliveryPayload, cb: (err?: Error | null) => void): void
+}
+
+export interface SessionDelegate<T extends Config = Config> {
+  startSession: (client: Client<T>, session: Session) => Client
+  pauseSession: (client: Client<T>) => void
+  resumeSession: (client: Client<T>) => Client
+}
+
+export default class Client<T extends Config = Config> {
+  private readonly _notifier?: Notifier
+  // This ought to be Required<T> but the current version of TypeScript doesn't seem to like it
+  public readonly _config: Required<Config>
+  private readonly _schema: any
+
+  public _delivery: Delivery
+
+  private readonly _plugins: { [key: string]: any }
+
+  public _breadcrumbs: Breadcrumb[]
+  public _session: Session | null
+  public _metadata: { [key: string]: any }
+  public _featuresIndex: { [key: string]: number }
+  public _features: Array<FeatureFlag | null>
+
+  private _context: string | undefined
+
+  public _user: User
+
+  public _logger: LoggerConfig
+
+  public readonly _cbs: {
+    e: OnErrorCallback[]
+    s: OnSessionCallback[]
+    sp: any[]
+    b: OnBreadcrumbCallback[]
+  }
+
+  private readonly Client: typeof Client
+  private readonly Event: typeof Event
+  private readonly Breadcrumb: typeof Breadcrumb
+  private readonly Session: typeof Session
```

Comment:
I've generally started out with things being `private readonly` and loosened as required. In the next PR where the generated type defs are used instead of the manually created ones even more of this properties become public